Create 2025-05-07-solidity-0.8.30-release-announcement.md (#143)
ethereum/solidity-website: src/posts/2025-05-07-solidity-0.8.30-release-announcement.md

This is meant to offset the bandwidth increase caused by another included EIP ([EIP-7691](https://eips.ethereum.org/EIPS/eip-7691)) through reducing the maximum block size and its variance, without negatively impacting regular users.

The increase in calldata cost could affect the validity of decisions that compiler components have to make based on it.
Most notably, the `ConstantOptimizer`, which is a part of the [opcode-based optimizer module](https://docs.soliditylang.org/en/v0.8.30/internals/optimizer.html#opcode-based-optimizer-module).
To illustrate the issue, let's consider the constant `2**256 - 1`.
It could be encoded either as `PUSH32 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff` (3 gas, 33 bytes) or `PUSH0 NOT` (5 gas, 2 bytes).
The optimizer has to decide which one is a better choice and that involves making assumptions about the cost of each byte.
Since the contract initcode is passed as calldata into the creation transaction, the cost of calldata is a part of that calculation, and has some influence on which form is chosen.

Despite this, we have decided not to introduce any adjustments to the calculations performed by the compiler.

> Firstly, it is nearly impossible to do so, due to fact that the cost increase is non-local and cannot easily be accounted for on the level of individual opcodes,

Reviewer: ```suggestion
Firstly, it is nearly impossible to do so, due to fact that the cost increase is non-local and cannot easily be accounted for on the level of individual opcodes.
```

The final cost now depends on the execution cost of the transaction, which is variable and depends on the input, the state and comes from the entirety of the path taken through the code.
Secondly, going back to the `ConstantOptimizer` example, it is better mitigated by user's choice of optimization parameters, specifically `--optimizer-runs`.
The default of `200` results in the compiler attempting to strike a balance between the cost of contract deployment and the cost of all expected calls over its lifetime and the higher calldata cost means the former may be underestimated.
However, it is a common practice to choose either a very low or very high value to explicitly prioritize one or the other and this effectively removes the calldata cost from consideration.

### EIP-7702: Set EOA account code

> EIP-7702 in Pectra adds a new transaction type that allows an EOF owner to delegate their account to an existing contract and this way make it act as one.

Reviewer: ```suggestion
EIP-7702 in Pectra adds a new transaction type that allows an EOA owner to delegate their account to an existing contract and this way make it act as one.
```

This is an important step on the road to full account abstraction.
Read more [in the EIP](https://eips.ethereum.org/EIPS/eip-7702)!

This newest release does not introduce any additional changes related to the above EIP.
However, please note that the previous release of the compiler ([v0.8.29](https://soliditylang.org/blog/2025/03/12/solidity-0.8.29-release-announcement)) introduced control over storage layout specifically to allow multiple delegates to safely share storage belonging to the same EOA.

### EIP-2537: Precompile for BLS12-381 curve operations

> [EIP-2537](https://eips.ethereum.org/EIPS/eip-2537) adds operation on BLS12-381 curve as a precompile in a set necessary to efficiently perform operations such as BLS signature verification.

Reviewer: We could expand on this a bit:
```suggestion
[EIP-2537](https://eips.ethereum.org/EIPS/eip-2537), nearly 5 years in the making, introduces a precompile necessary to perform BLS signature verification on chain.
While Ethereum transactions use the much faster ECDSA signatures, BLS signatures are already used in the consensus and have long been requested due to a very desirable property: they can be aggregated.
```

> The current release does not introduce any global builtins for the precompiles and thus, users will have to make calls to them manually. We understand that there might be community interest in support for BLS signatures in Solidity. While we are open to discussion and would love to understand how a feature like this could be useful and can be implemented, we currently have no plans to implement it in the language.

Reviewer 2: ```suggestion
Since builtins are essentially function calls that are straightforward to implement on top of the existing language as part of an external library, we believe that a feature like this is better delegated to a standard library rather supported directly by the Solidity compiler.

The current release does not introduce any global builtins for the precompiles. We understand that there might be community interest in support for BLS signatures in Solidity. While we are open to discussion and would love to understand how a feature like this could be useful and can be implemented, we currently have no plans to implement it in the language.
```

I feel like starting with this paragraph gives a better support to then declare that we won't implement global builtins.

Author: I agree. It does read better. :)

> The current release does not introduce any global builtins for the precompiles and thus, users will have to make calls to them manually. We understand that there might be community interest in support for BLS signatures in Solidity. While we are open to discussion and would love to understand how a feature like this could be useful and can be implemented, we currently have no plans to implement it in the language.

Reviewer 2: ```suggestion
The current release does not introduce any global builtins for the precompiles and thus, users will have to make calls to them manually. We understand that there might be community interest in support for BLS signatures in Solidity. While we are open to discussion and would love to understand the use cases for it, we currently have no plans to implement it in the language.
```

However, please note that the previous release of the compiler ([v0.8.29](https://soliditylang.org/blog/2025/03/12/solidity-0.8.29-release-announcement)) introduced control over storage layout specifically to allow multiple delegates to safely share storage belonging to the same EOA.

### EIP-2537: Precompile for BLS12-381 curve operations
[EIP-2537](https://eips.ethereum.org/EIPS/eip-2537), nearly 5 years in the making, introduces a precompile necessary to perform BLS signature verification on chain.
While Ethereum transactions use the much faster ECDSA signatures, BLS signatures are already used in the consensus and have long been requested due to a very desirable property: they can be aggregated.

The current release does not introduce any global builtins for the precompiles and thus, users will have to make calls to them manually. We understand that there might be community interest in support for BLS signatures in Solidity. While we are open to discussion and would love to understand how a feature like this could be useful and can be implemented, we currently have no plans to implement it in the language.

> Since builtins are essentially function calls that are straightforward to implement on top of the existing language as part of an external library, we do believe that a feature like this is better delegated to a standard library rather supported directly by the Solidity compiler.

Reviewer: How about this?
```suggestion
Since precompile builtins are essentially thin wrappers over external function calls, they are straightforward to implement using existing features and do not require extending the language.
We believe that a feature like this is better delegated to an external library rather than being hard-coded in the Solidity compiler, though we do not exclude the possibility of it eventually becoming a part of the planned standard library.
```

This is meant to offset the bandwidth increase caused by another included EIP ([EIP-7691](https://eips.ethereum.org/EIPS/eip-7691)) through reducing the maximum block size and its variance, without negatively impacting regular users.

The increase in calldata cost could affect the validity of decisions that compiler components have to make based on it.
Most notably, the `ConstantOptimizer`, which is a part of the [opcode-based optimizer module](https://docs.soliditylang.org/en/v0.8.30/internals/optimizer.html#opcode-based-optimizer-module).
To illustrate the issue, let's consider the constant `2**256 - 1`.
It could be encoded either as `PUSH32 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff` (3 gas, 33 bytes) or `PUSH0 NOT` (5 gas, 2 bytes).
The optimizer has to decide which one is a better choice and that involves making assumptions about the cost of each byte.
Since the contract initcode is passed as calldata into the creation transaction, the cost of calldata is a part of that calculation, and has some influence on which form is chosen.

Despite this, we have decided not to introduce any adjustments to the calculations performed by the compiler.

> Firstly, it is nearly impossible to do so, due to fact that the cost increase is non-local and cannot easily be accounted for on the level of individual opcodes.

Reviewer: I forgot a `the` here. Probably not worth another PR just to fix it though.

```suggestion
Firstly, it is nearly impossible to do so, due to the fact that the cost increase is non-local and cannot easily be accounted for on the level of individual opcodes.
```
